cpp/src/io/parquet/writer_impl.cu

```diff
@@ -568,27 +567,24 @@ struct leaf_schema_fn {
   template <typename T>
   std::enable_if_t<cudf::is_fixed_point<T>(), void> operator()()
   {
-    // If writing arrow schema, then convert d32 and d64 to d128
-    if (write_arrow_schema or std::is_same_v<T, numeric::decimal128>) {
+    if constexpr (std::is_same_v<T, numeric::decimal32>) {
+      col_schema.type              = Type::INT32;
+      col_schema.stats_dtype       = statistics_dtype::dtype_int32;
+      col_schema.decimal_precision = MAX_DECIMAL32_PRECISION;
+      col_schema.logical_type      = LogicalType{DecimalType{0, MAX_DECIMAL32_PRECISION}};
+    } else if constexpr (std::is_same_v<T, numeric::decimal64>) {
+      col_schema.type              = Type::INT64;
+      col_schema.stats_dtype       = statistics_dtype::dtype_decimal64;
+      col_schema.decimal_precision = MAX_DECIMAL64_PRECISION;
+      col_schema.logical_type      = LogicalType{DecimalType{0, MAX_DECIMAL64_PRECISION}};
+    } else if constexpr (std::is_same_v<T, numeric::decimal128>) {
       col_schema.type              = Type::FIXED_LEN_BYTE_ARRAY;
       col_schema.type_length       = sizeof(__int128_t);
```

Comment:
Is type_length inferred for the other types?

Reply:
Not needed for smaller types since they are represented as `Type::INTXX` instead of `Type::FLBA` for `decimal128`